src/lib.rs

```diff
@@ -0,0 +1,718 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+use borsh::schema::{BorshSchemaContainer, Declaration, Definition, Fields, VariantName};
+use schemars::schema::{RootSchema, Schema};
+use serde::{Deserialize, Serialize};
+
+/// Current version of the ABI schema format.
+const ABI_SCHEMA_SEMVER: &str = env!("CARGO_PKG_VERSION");
+
+/// Contract ABI.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct AbiRoot {
+    /// Semver of the ABI schema format.
+    abi_schema_version: String,
+    /// Metadata information about the contract.
+    pub metadata: AbiMetadata,
+    /// Core ABI information (functions and types).
+    pub abi: AbiEntry,
+}
+
+impl AbiRoot {
+    #[cfg(feature = "versioned-entries")]
+    pub fn new(metadata: AbiMetadata, versioned_abi: VersionedAbiEntry) -> AbiRoot {
+        AbiRoot {
+            abi_schema_version: versioned_abi.abi_schema_version,
+            metadata,
+            abi: versioned_abi.abi,
+        }
+    }
+
+    pub fn schema_version(&self) -> &str {
+        &self.abi_schema_version
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Default)]
+pub struct AbiMetadata {
+    /// The name of the smart contract.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    /// The version of the smart contract.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
+    /// The authors of the smart contract.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub authors: Vec<String>,
+    /// Other arbitrary metadata.
+    #[serde(default, flatten, skip_serializing_if = "HashMap::is_empty")]
+    pub other: HashMap<String, String>,
+}
+
+/// Core ABI information.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct AbiEntry {
+    /// ABIs of all contract's functions.
+    pub functions: Vec<AbiFunction>,
+    /// Root JSON Schema containing all types referenced in the functions.
+    pub root_schema: RootSchema,
+}
+
+/// Core ABI information, with schema version.
+#[cfg(feature = "versioned-entries")]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct VersionedAbiEntry {
+    /// Semver of the ABI schema format.
+    abi_schema_version: String,
+    #[serde(flatten)]
+    pub abi: AbiEntry,
+}
+
+#[cfg(feature = "versioned-entries")]
```

Comment:
Also, any reason for this to be a feature?

Reply:
Yeah, going with the current schema definition. I thought we could treat `VersionedAbiEntry` as a private type. Only used by the code-generated contract describer and `cargo-near`. A `VersionedAbiEntry` is effectively an `AbiRoot` without `metadata`, not needed in the public API.